Updates discounts in the staff dashboard to reflect the current state of the art

jkachel opened the pull request:
#### Pre-Flight checklist

- [X] Screenshots and design review for any changes that affect layout or styling
  - [X] Desktop screenshots
- [X] Testing
  - [X] Code is tested
  - [X] Changes have been manually tested

#### What are the relevant tickets?

Closes #1266

#### What's this PR do?

Updates the Discounts section of the staff dashboard app to be more in line with the current state of the art for the system:
- Dashboard
   - Adds search functionality
   - Updates the Flexible Price Tier tag to read Financial Assistance Tier
   - Adds the activation and expiration dates
   - Adds a Redeemed tag if the discount is single-use and has been used
- Create/Edit
   - Refactored out the form into a separate reusable component
   - Added maximum redemption field (hidden unless the discount is unlimited use)
   - Added fields for activation and expiry
   - Added financial assistance tier field (defaults to not)
- Edit
   - Adds tables to manage assigned users and products for the discount
- Show
   - Added additional fields to the top, in a two-column format: max redemptions (if applicable) and activation/expiration dates  
   - Added the Redeemed tag for single-use discounts that have been redeemed
   - Added a table to show the tiers the discount is used for (if it's a financial assistance discount)
   - Moved the Redemptions table up just under the discount information
- Meta
   - Fixed the route for `/staff-dashboard` so that deep links into the dashboard will actually work now 
   - Added a serializer for the financial assistance tiers
   - Added endpoints for tiers for a given discount 

#### How should this be manually tested?

Automated tests should pass. 

Discount Dashboard:
- By default, the list should show all discounts in the system with working pagination controls.
- Active From and Expires should be set if the discount is time-limited. 
- Discounts should be tagged "Financial Assistance Tier" if the discount is for financial assistance.
- Discounts should be tagged "Redeemed" if they're one-time or one-time-per-user and the discount has been used.
- Search should work, and search items should be ANDed.

Create:
- Should work.
- The form's Financial Assistance field should default to Regular Discount.
- The activation and expiration date fields should default to null.
- If the discount is set to Unlimited, you should be able to specify a Max Redemptions value.
- You should not see a Products or User Assignments table.

Edit:
- Should work.
- The Financial Assistance, Activation Date, and Expiration Date fields should all populate properly. 
- Max Redemptions should appear if the discount is Unlimited use and should populate properly.
- You should see the Products table.
   - Products should display any products the discount has been assigned to.
   - Clicking Delete should remove the product from the discount. 
   - Clicking Add Product should open a modal with a search field.
   - You should be able to search for products based on product name or readable ID. (Note that most products use the courseware readable ID as the product name so you may want to add some products that don't follow that convention to test with.) 
   - Clicking Add should add products to the list, and they should show up in the underlying table. (You should be able to see this if the Products table isn't obscured.) 
   - Clicking Close should close the modal.
- You should see the Assigned Users table.
   - The table should show any users assigned to the discount.
   - Operations should work in the same manner as the Products table.
   - The Assign Users modal should include some text describing how the user discounts work. 

View:
- The discount should display the Redeemed tag if the discount is one-time or one-time-per-user and has been used.
- If the discount has an activation or expiration date, it should be displayed on the right.
- If the discount has a usage limit, it should be displayed on the right.  
- The Redemptions, Assigned Users, and Products tables should all work as expected.
- Assigned Users and Products should _not_ display the Add Products or Assign Users buttons. 
- The Financial Assistance Tiers table should be visible and should show any tiers that are using the discount. 
- The tables should be in the order of Redemptions, Financial Assistance Tiers, Assigned Users, Products. 

To test the URL change: the easiest way to do this is just to go to a page within the staff dashboard (any of them) and hit reload. The route for the staff dashboard only matched on `^/staff-dashboard` so that would fail if you were going to a specific resource within it (for example, `/staff-dashboard/discounts`) - the new regex should fix the issue.


#### Screenshots (if appropriate)

Dashboard: 
![image](https://user-images.githubusercontent.com/945611/210883322-3b9427fa-c7d9-4b6c-b50b-5fc634c2f9ac.png)

Create:
![image](https://user-images.githubusercontent.com/945611/211878151-6174cf26-acd5-4fa5-a486-967784ae3d7f.png)


Edit:
![image](https://user-images.githubusercontent.com/945611/211877837-f3df47ec-16a7-446e-9c87-1d9776dff373.png)

Edit - Assign Products dialog (with search results):
![image](https://user-images.githubusercontent.com/945611/211877944-c65e7425-8536-4caf-ad15-f676977f477f.png)

Edit - Assign Users dialog (with search results):
![image](https://user-images.githubusercontent.com/945611/211878024-62840bcf-766c-4f11-bf75-c065156b6998.png)

Show:
![image](https://user-images.githubusercontent.com/945611/210883552-aca96820-f5e9-441d-87c2-c3789f796e36.png)

Show (w/ redemptions, assigned users, assigned products):
![Screenshot 2023-01-05 at 3 52 25 PM](https://user-images.githubusercontent.com/945611/210887080-6770d512-07cb-4aa4-8322-7aad152feb7d.png)

asadali145 commented on users/views.py:
```diff
@@ -88,3 +92,25 @@ def list(self, request):  # pylint:disable=unused-argument
+    permission_classes = [IsAdminUser]
+    pagination_class = RefinePagination
+
+    def get_queryset(self):
+        name_search = self.request.query_params.get("q")
+
+        if name_search is None:
+            return User.objects.all()
+
+        return User.objects.filter(
+            Q(username__icontains=name_search)
+            | Q(name__icontains=name_search)
+            | Q(email__icontains=name_search)
+        ).all()
```
We can use [SearchFilter](https://www.django-rest-framework.org/api-guide/filtering/#searchfilter).

asadali145 commented on frontend/staff-dashboard/src/pages/discounts/show.tsx:
```diff
@@ -16,26 +14,49 @@ export const DiscountShow = () => {
-                { record?.for_flexible_pricing ? (<Tag color="green">Flexible Pricing Tier</Tag>) : null }</Space>
+                <Space>
+                    Discount Detail
+                    { record?.for_flexible_pricing ? (<Tag color="green">Flexible Pricing Tier</Tag>) : null }
```
Maybe we should change this to `Financial Assistance Tier` as well.

asadali145 reviewed (commented):
Awesome Work! Just a couple of recommendations.

jkachel commented on ecommerce/models.py:
```diff
@@ -256,6 +256,10 @@ class Discount(TimestampedModel):
 
+    @cached_property
+    def is_redeemed(self):
+        return DiscountRedemption.objects.filter(redeemed_discount=self).exists()
```
Hmm. We should probably clear the discount redemptions in this case. This is outside the scope of this PR, though.

jkachel commented:
Last commit(s) rebase from main and address the comments:
- Fixed the tag 
- Delete ops are compressed down into a single line
- Refactored to use SearchFilter or Django Filters

asadali145 commented on frontend/staff-dashboard/src/pages/discounts/show.tsx:
```diff
@@ -16,26 +14,49 @@ export const DiscountShow = () => {
+            {record ? <RedemptionList record={record} key={`redemptions-${record?.id}`} /> : null}
 
-            {record ? <UserAssignments record={record} key={record?.user} /> : null}
+            {record ? <FinAidTiers record={record} key={`tiers-${record?.id}`} /> : null}
```
We could skip this section if record.for_flexinle_pricing is true.

asadali145 commented on frontend/staff-dashboard/src/pages/discounts/show.tsx:
```diff
@@ -16,26 +14,49 @@ export const DiscountShow = () => {
 
-            <Title level={5}>Amount</Title>
-            <Text>{record?.amount} {record?.discount_type}</Text>
+            {record ? <RedemptionList record={record} key={`redemptions-${record?.id}`} /> : null}
```
It would be helpful to update `RedemptionList` to include the user email and some order details.

asadali145 reviewed (commented):
Apart from a few minor changes. Everything looks great and a fine refactoring effort for the discounts dashboard.

jkachel commented:
https://github.com/mitodl/mitxonline/pull/1324/commits/c6834afa5ad0a5f883643aa78efef4cb4b1d92c4 should have the last batch of things worked out.

asadali145 commented on ecommerce/views/v0/__init__.py:
```diff
@@ -2,97 +2,127 @@
+    permission_classes = [IsAdminUser]
 
+    def get_queryset(self):
+        name_search = self.request.query_params.get("q")
```
We need to update the query param name, We are using `search` at the frontend so this search is not working while assigning products.

asadali145 reviewed (approved):
Apart from a single issue that I mentioned, Everything LGTM! Awesome Job! You can go ahead and merge after fixing that.
